crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3327,7 +3339,13 @@ impl<'db> TypeInferenceBuilder<'db> {
             }
         }
 
-        self.infer_expression(target);
+        // Annotated assignments to non-names are not definitions, so we can only be here
+        // if the target is a name. In this case, we can simply store the declared type in
+        // the store-context name node instead of calling `infer_expression`, which would
+        // be more work and may lead to a different type (the inferred type of the right-
+        // hand side, or `Never` if there is no right-hand side).
+        debug_assert!(target.is_name_expr());
+        self.store_expression_type(target, declared_ty.inner_type());
```

Comment:
This (and a similar change for other annotated assignments above) might be a bit controversial. It's related to https://github.com/astral-sh/ruff/issues/17239 in that it's a matter of showing the declared type vs the inferred type. If there is no right hand side, I think this is clearly an improvement (over showing `Never`). But if there is (`x: int = 1`), we now show `int` when hovering over `x`, whereas we would previously have shown `Literal[1]`).

Reply:
> Should we change this to show the right-hand side value type for now? (if there is one)

That feels like it makes sense to me? I can see arguments both ways, though; I don't have a strong opinion

Reply:
Ok, we're using the inferred type now.